### PR TITLE
Feat/synapse

### PR DESCRIPTION
**Pull request by dextracker**
## Description

Adds support for synapse protocol swap to:

- Mainnet
  -`DAI/USDC/USDT` Stableswap pool
- BSC
  -`nUSD/USDC/BUSD/USDT` Stableswap pool 
- Polygon
  -`nUSD/USDC/DAI/USDT` Stableswap pool  
- Fantom
  -`nUSD/USDC/MIM/USDT` Stableswap pool 
  -`nETH/ETH` ETH pool 
- Avalanche (`nETH/avWETH` pool not working atm)
  -`nUSD/USDC/DAI/USDT` Stableswap pool 
 -```(`nETH/avWETH` pool not working atm)```
- Optimism
  -`nETH/WETH` ETH pool
    -```(wasnt able to check this one with simbot because there isnt a wallet that we can grab from but quotes are working)```

## Checklist:

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: packages/asset-swapper/src/utils/market_operation_utils/constants.ts [hunk 1]
```diff
@@ -462,6 +472,9 @@ export const MAINNET_TOKENS = {
     CRV: '0xd533a949740bb3306d119cc777fa900ba034cd52',
     MIM: '0x99d8a9c45b2eca8864373a26d1459e3dff1e17f3',
     EURT: '0xc581b735a1688071a1746c968e0798d642ede491',
+    // Synapse ecosystem
+    nUSD: '0x1B84765dE8B7566e4cEAF4D0fD3c5aF52D3DdE4F',
```
Comment (shawnxin89) [inline packages/asset-swapper/src/utils/market_operation_utils/constants.ts:476]: Should we convert it to lowercase address?

Comment (dextracker) [reply packages/asset-swapper/src/utils/market_operation_utils/constants.ts]: thank you! some of the pools were uppercase as well

File: packages/asset-swapper/src/utils/market_operation_utils/constants.ts [hunk 2]
```diff
@@ -663,6 +688,37 @@ export const NERVE_POOLS = {
     threePool: '0x1b3771a66ee31180906972580ade9b81afc5fcdc',
 };
 
+export const SYNAPSE_MAINNET_POOLS = {
+    nUSDLP: '0x1116898DdA4015eD8dDefb84b6e8Bc24528Af2d8',
```
Comment (shawnxin89) [inline packages/asset-swapper/src/utils/market_operation_utils/constants.ts]: Convert to lowercase?

File: packages/asset-swapper/src/utils/market_operation_utils/constants.ts [hunk 3]
```diff
@@ -1343,6 +1423,87 @@ export const NERVE_BSC_INFOS: { [name: string]: CurveInfo } = {
     },
 };
 
+export const SYNAPSE_BSC_INFOS: { [name: string]: CurveInfo } = {
+    [SYNAPSE_BSC_POOLS.nUSDLP]: {
+        exchangeFunctionSelector: CurveFunctionSelectors.swap,
+        sellQuoteFunctionSelector: CurveFunctionSelectors.calculateSwap,
+        buyQuoteFunctionSelector: CurveFunctionSelectors.None,
+        poolAddress: SYNAPSE_BSC_POOLS.nUSDLP,
+        tokens: [BSC_TOKENS.nUSD, BSC_TOKENS.BUSD, BSC_TOKENS.USDT, BSC_TOKENS.USDC],
```
Comment (shawnxin89) [inline packages/asset-swapper/src/utils/market_operation_utils/constants.ts:1429]: I'm not sure whether the order of the tokens matters here (it matters for curve pools afair). The order for this pool is nUSD, BUSD, USDC, USDT on Synapse website.

File: packages/asset-swapper/src/utils/market_operation_utils/constants.ts [hunk 4]
```diff
@@ -188,13 +192,14 @@ export const SELL_SOURCE_FILTER_BY_CHAIN_ID = valueByChainId<SourceFilters>(
...
             ERC20BridgeSource.SushiSwap,
+            ERC20BridgeSource.Synapse,
         ]),
         [ChainId.Celo]: new SourceFilters([
             ERC20BridgeSource.UbeSwap,
             ERC20BridgeSource.SushiSwap,
             ERC20BridgeSource.MultiHop,
         ]),
-        [ChainId.Optimism]: new SourceFilters([ERC20BridgeSource.UniswapV3]),
+        [ChainId.Optimism]: new SourceFilters([ERC20BridgeSource.UniswapV3, ERC20BridgeSource.Synapse]),
```
Comment (dekz) [inline packages/asset-swapper/src/utils/market_operation_utils/constants.ts:202]: Can you also drop in `ERC20BridgeSource.MultiHop` here
```suggestion
        [ChainId.Optimism]: new SourceFilters([
            ERC20BridgeSource.UniswapV3,
            ERC20BridgeSource.Synapse,
            ERC20BridgeSource.MultiHop,
         ]),
```

File: packages/asset-swapper/src/utils/market_operation_utils/constants.ts [hunk 5]
```diff
@@ -321,13 +330,14 @@ export const BUY_SOURCE_FILTER_BY_CHAIN_ID = valueByChainId<SourceFilters>(
...
```
Comment (dekz) [inline packages/asset-swapper/src/utils/market_operation_utils/constants.ts:340]:
```suggestion
        [ChainId.Optimism]: new SourceFilters([
            ERC20BridgeSource.UniswapV3,
            ERC20BridgeSource.Synapse,
            ERC20BridgeSource.MultiHop,
         ]),
```

**Review by dekz — CHANGES_REQUESTED** [review]

**Review by shawnxin89 — APPROVED** [review]
Looks really nice

**Pull request merged**